extensions/amp-form/0.1/amp-form.js

```diff
@@ -260,17 +316,114 @@ function onInvalidInputBlur_(event) {
+}
 
-  input.addEventListener('keyup', onInvalidInputKeyUp_);
-  input.addEventListener('blur', onInvalidInputBlur_);
-  validationBubble.show(input, input.validationMessage);
+
+/**
+ * Checks user validity for all inputs, fieldsets and the form.
+ * @param {!HTMLFormElement} form
+ * @return {boolean} Whether the form is currently valid or not.
+ */
+function checkUserValidityOnSubmission(form) {
+  const inputs = form.querySelectorAll('input,select,textarea,fieldset');
+  for (let i = 0; i < inputs.length; i++) {
+    checkUserValidity(inputs[i]);
+  }
+  return checkUserValidity(form);
+}
+
+
+/**
+ * Returns the user validity state of the element.
+ * @param {!HTMLInputElement|!HTMLFormElement|!HTMLFieldSetElement} element
+ * @return {string}
+ */
+function getUserValidityStateFor(element) {
+  if (element.classList.contains('user-valid')) {
+    return UserValidityState.USER_VALID;
+  } else if (element.classList.contains('user-invalid')) {
+    return UserValidityState.USER_INVALID;
+  }
+
+  return UserValidityState.NONE;
+}
+
+
+/**
+ * Checks user validity which applies .user-valid and .user-invalid AFTER the user
+ * interacts with the input by moving away from the input (blur) or by changing its
+ * value (input).
+ *
+ * See :user-invalid spec for more details:
+ *   https://drafts.csswg.org/selectors-4/#user-pseudos
+ *
+ * The specs are still not fully specified. The current solution tries to follow a common
+ * sense approach for when to apply these classes. As the specs gets clearer, we should
+ * strive to match it as much as possible.
+ *
+ * @param {!HTMLInputElement|!HTMLFormElement|!HTMLFieldSetElement} element
+ * @param {boolean=} propagate Whether to propagate the user validity to ancestors.
+ * @returns {boolean} Whether the element is valid or not.
+ */
+function checkUserValidity(element, propagate = false) {
+  let shouldPropagate = false;
+  const previousValidityState = getUserValidityStateFor(element);
+  const isCurrentlyValid = element.checkValidity && element.checkValidity();
+  if (previousValidityState != UserValidityState.USER_VALID &&
+      isCurrentlyValid) {
+    element.classList.add('user-valid');
+    element.classList.remove('user-invalid');
+    // Don't propagate user-valid unless it was marked invalid before.
+    shouldPropagate = previousValidityState == UserValidityState.USER_INVALID;
+  } else if (previousValidityState != UserValidityState.USER_INVALID &&
+      !isCurrentlyValid) {
+    element.classList.add('user-invalid');
+    element.classList.remove('user-valid');
+    // Always propagate an invalid state change. One invalid input field is
+    // guaranteed to make the fieldset and form invalid as well.
+    shouldPropagate = true;
+  }
+
+  if (propagate && shouldPropagate) {
+    // Propagate user validity to ancestor fieldsets.
+    const ancestors = ancestorElementsByTag(element, 'fieldset');
+    for (let i = 0; i < ancestors.length; i++) {
+      checkUserValidity(ancestors[i]);
```

Comment:
Does this work correctly assuming multiple inputs in one fieldset: e.g. one invalid and one valid?